src/Microsoft.Framework.Runtime/FileGlobbing/PatternsGroup.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            IncludeLiterals = includeLiterals;
+            IncludePatterns = includePatterns;
+            ExcludePatterns = excludePatterns;
+
+            _matcher.AddIncludePatterns(IncludePatterns);
+            _matcher.AddExcludePatterns(ExcludePatterns);
+        }
+
+        public static PatternsGroup Build(JObject rawProject, string projectDirectory, string name, string legacyName, IEnumerable<string> fallback, IEnumerable<string> buildInExcludePatterns, bool includePatternsOnly = false)
+        {
+            var token = rawProject[name];
+            if (legacyName != null)
+            {
+                var legacyToken = rawProject[legacyName];
+                if (legacyToken != null && token == null)
+                {
+                    token = legacyToken;
+                }
+            }
+
+            var includePatterns = PatternsCollectionHelper.GetPatternsCollection(token, projectDirectory, fallback);
+
+            if (includePatternsOnly)
+            {
+                return new PatternsGroup(includePatterns);
+            }
+
+            var excludePatterns = PatternsCollectionHelper.GetPatternsCollection(rawProject, projectDirectory, name + "Exclude")
+                                                          .Concat(buildInExcludePatterns)
+                                                          .Distinct();
+
+            var includeLiterals = PatternsCollectionHelper.GetPatternsCollection(rawProject, projectDirectory, name + "Files")
+                                                          .Distinct();
+
+            return new PatternsGroup(includePatterns, excludePatterns, includeLiterals);
+        }
+
+        public IEnumerable<string> IncludeLiterals { get; }
+
+        public IEnumerable<string> IncludePatterns { get; }
+
+        public IEnumerable<string> ExcludePatterns { get; }
+
+        public IEnumerable<PatternsGroup> ExcludePatternsGroup { get { return _excludeGroups; } }
+
+        public PatternsGroup ExcludeGroup(PatternsGroup group)
+        {
+            _excludeGroups.Add(group);
+
+            return this;
+        }
+
+        public IEnumerable<string> SearchFiles(string rootPath)
+        {
+            // literal included files are added at the last, but the search happens early
+            // so as to make the process fail early in case there is missing file. fail early
+            // helps to avoid unnecessary globing for performance optimization
+            var literalIncludedFiles = new List<string>();
+            foreach (var literalRelativePath in IncludeLiterals)
+            {
+                var fullPath = Path.GetFullPath(Path.Combine(rootPath, literalRelativePath));
+                if (!File.Exists(fullPath))
+                {
+                    throw new InvalidOperationException(string.Format("Can't find file {0}", literalRelativePath));
+                }
+
+                literalIncludedFiles.Add(fullPath.Replace('\\', '/'));
```

Comment:
Elsewhere in XRE we use Path.DirectorySeparatorChar and Path.AltDirectorySeparatorChar